### PR TITLE
refactor/ Binance websockets removal

### PR DESCRIPTION
**Pull request by CrimsonJacket**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Refactor `websockets` package in Binance to use `aiohttp` websockets instead.

**Tests performed by the developer**:
Refactored `BinanceAPIOrderBookDataSource` and `BinanceAPIUserStreamDataSource` to use `aiohttp` websockets.
Added unittests for new websockets 
Modified some unit tests to use `aioresponses`

**Tips for QA testing**:
Ensure that Binance and Binance_US Spot connector works as usual

### CODE REVIEW

**Review by petioptrv — CHANGES_REQUESTED** [review]
Just one comment in the code, and the coverage. The rest looks good to me

File: test/hummingbot/connector/exchange/binance/test_binance_user_stream_data_source.py [hunk 1]
```diff
@@ -177,3 +240,30 @@ def test_listen_for_user_stream_no_listen_key(self, mock_api, mock_next_tick, mo
...
+        mock_ws.return_value.receive.side_effect = lambda: self._raise_exception(Exception("TEST ERROR"))
+        mock_ws.close.return_value = None
+
+        self.listening_task = self.ev_loop.create_task(
+            self.data_source.listen_for_user_stream(self.ev_loop, msg_queue)
+        )
+
+        with self.assertRaises(asyncio.TimeoutError):
+            self.ev_loop.run_until_complete(
+                asyncio.wait_for(self.listening_task, 1)
```
Comment (petioptrv) [inline test/hummingbot/connector/exchange/binance/test_binance_user_stream_data_source.py:266]: For this one, we can use the same approach as `NdaxExchangeTests._create_exception_and_unlock_test_with_event` to shave off the second

**Review by petioptrv — CHANGES_REQUESTED** [review]
Left some comments for the tests that wait asynchronously 1 second to try and instead use event-setting. Reason is that currently the CD pipeline, runs around 950 tests in about 84 seconds in the "Run stable tests and calculate coverage" step. So waiting for 1 second has a big relative impact

**Review by petioptrv — DISMISSED** [review]
Changes LGTM!

File: hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py [hunk 2]
```diff
@@ -152,64 +145,91 @@ async def get_new_order_book(self, trading_pair: str) -> OrderBook:
...
+                    if "result" in json_msg:
+                        continue
+                    trade_msg: OrderBookMessage = BinanceOrderBook.trade_message_from_exchange(json_msg)
                     output.put_nowait(trade_msg)
             except asyncio.CancelledError:
                 raise
             except Exception:
                 self.logger().error("Unexpected error with WebSocket connection. Retrying after 30 seconds...",
                                     exc_info=True)
+                ws and await ws.close()
```
Comment (aarmoa) [inline hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py:202]: This line should be inside a `finally` statement, not only in the `except Exception`

File: hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py [hunk 3]
```diff
@@ -152,64 +145,91 @@ async def get_new_order_book(self, trading_pair: str) -> OrderBook:
...
                     order_book_message: OrderBookMessage = BinanceOrderBook.diff_message_from_exchange(
-                        msg, time.time())
+                        json_msg, time.time())
                     output.put_nowait(order_book_message)
             except asyncio.CancelledError:
                 raise
             except Exception:
                 self.logger().error("Unexpected error with WebSocket connection. Retrying after 30 seconds...",
                                     exc_info=True)
+                ws and await ws.close()
```
Comment (aarmoa) [inline hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py:232]: This line should be inside a `finally` statement, not only in the `except Exception`

File: hummingbot/connector/exchange/binance/binance_api_user_stream_data_source.py [hunk 4]
```diff
@@ -62,57 +61,58 @@ async def get_listen_key(self):
...
+    async def _create_websocket_connection(self) -> aiohttp.ClientWebSocketResponse:
+        """
+        Initialize WebSocket client for UserStreamDataSource
+        """
+        try:
+            url = f"{CONSTANTS.WSS_URL.format(self._domain)}/{self._current_listen_key}"
+            self.logger().info(f"Connecting to {url}.")
+            return await aiohttp.ClientSession().ws_connect(url=url,
+                                                            heartbeat=30.0,
+                                                            autoping=False)
```
Comment (aarmoa) [inline hummingbot/connector/exchange/binance/binance_api_user_stream_data_source.py:94]: Is it correct that the autoping for the user stream websocket is disabled but enabled for the order book stream websocket? What is the heartbeat used for if autoping is disabled?

File: hummingbot/connector/exchange/binance/binance_api_user_stream_data_source.py [hunk 5]
```diff
@@ -136,33 +136,31 @@ async def _manage_listen_key_task_loop(self):
...
+                await self._listen_key_initialized_event.wait()
+
+                self._ws = await self._create_websocket_connection()
+
+                async for msg in self._iter_messages(self._ws):
+                    if msg.type == aiohttp.WSMsgType.PING:
+                        self.logger().debug("Received PING frame. Sending PONG frame...")
+                        await self._ws.pong()
+                        continue
+                    output.put_nowait(ujson.loads(msg.data))
```
Comment (aarmoa) [inline hummingbot/connector/exchange/binance/binance_api_user_stream_data_source.py:153]:
```suggestion
                    output.put_nowait(msg.json())
```

File: test/hummingbot/connector/exchange/binance/test_binance_api_order_book_data_source.py [hunk 6]
```diff
@@ -284,12 +328,31 @@ def test_listen_for_trades_logs_exception(self, mock_ws):
...
+        mock_ws.return_value.receive_json.side_effect = lambda: self._raise_exception(Exception("TEST ERROR"))
+        mock_ws.close.return_value = None
+
+        self.listening_task = self.ev_loop.create_task(
+            self.data_source.listen_for_trades(self.ev_loop, msg_queue)
+        )
+
+        with self.assertRaises(asyncio.TimeoutError):
+            self.ev_loop.run_until_complete(
+                asyncio.wait_for(self.listening_task, 1)
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/hummingbot/connector/exchange/binance/test_binance_api_order_book_data_source.py:364]: Don't we still need here to replace the 1 second time out with waiting for the event `self.resume_test_event` to be set? (if we use the function `_create_exception_and_unlock_test_with_event`)

**Review by aarmoa — CHANGES_REQUESTED** [review]
I have added some comments. Some changes needed on some comments just to validate a couple of things.

File: hummingbot/connector/exchange/binance/binance_api_user_stream_data_source.py (hunk 4 above)
Comment (CrimsonJacket) [reply hummingbot/connector/exchange/binance/binance_api_user_stream_data_source.py:94]: Heartbeat is used to send ping frames to the server to determine if the connection is still up. While, `autoping` responds with a pong frame upon receiving a ping frame.

**Comment by rxlxrxsx** [comment]
Getting this error when stopping the bot but can only be seen in the client log pane and not recorded in the logs. Happens both on binance and binance_us.

Screen Shot 2021-10-03 at 1 11 33 PM

Logs and config: [binance_logs_config.zip](https://github.com/CoinAlpha/hummingbot/files/7272845/binance_logs_config.zip)

**Review by RC-13 — APPROVED** [review]
Confirming that the last reported issue has been address by the recent commit. LGTM!

**Review by aarmoa — APPROVED** [review]
Changes look good to me

**Pull request merged**